### PR TITLE
OperatorAll - prevent multiple terminal events

### PR DESCRIPTION
**Pull request by davidmoten**
As per discussion in #4242, if an operator maps an `onNext` emission to an `onError` emission downstream then it needs be defensive about an `onCompleted` being sent from upstream even if upstream has been unsubscribed.

Includes a unit test that failed on the original code.

### CODE REVIEW

File: src/main/java/rx/internal/operators/OperatorAll.java [hunk 1]
```diff
@@ -61,7 +61,10 @@ public void onNext(T t) {
 
             @Override
             public void onError(Throwable e) {
-                child.onError(e);
+                if (!done) {
+                    done = true;
+                    child.onError(e);
+                }
```
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorAll.java:73]: Call `RxJavaHooks.onError` so no exception is lost.

Comment (vanniktech) [inline src/main/java/rx/internal/operators/OperatorAll.java:73]: Since you pointed it out in both PRs would not this also apply to some other places?

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorAll.java:73]: Yes, anyplace that would drop the exception due to state/protocol considerations should instead call `RxJavaHooks.onError`.

**Comment by davidmoten** [comment]
- Updated with `RxJavaHooks.onError` 
- checked `done` at start of `onNext` 
- added unit tests for `(onNext, onNext)` and `(onNext, onError)` cases

**Comment by akarnokd** [comment]
:+1:

**Pull request merged**